In [1]:
import numpy as np 
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
import json 
#!pip install geopy
#!pip install folium
from geopy.geocoders import Nominatim
import requests 
from pandas.io.json import json_normalize
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn.cluster import KMeans
from bs4 import BeautifulSoup
import xml
import folium 


In [2]:
url = requests.get('https://en.wikipedia.org/w/index.php?title=List_of_postal_codes_of_Canada:_M&oldid=995657573').text
soup = BeautifulSoup(url,'lxml')

In [3]:
table_post = soup.find('table')
fields = table_post.find_all('td')

postcode = []
borough = []
neighbourhood = []

for i in range(0, len(fields), 3):
    postcode.append(fields[i].text.strip())
    borough.append(fields[i+1].text.strip())
    neighbourhood.append(fields[i+2].text.strip())
        
df = pd.DataFrame(data=[postcode, borough, neighbourhood]).transpose()
df.columns = ['Postcode', 'Borough', 'Neighbourhood']
df.head()

Postcode           Borough              Neighbourhood
0      M1A      Not assigned               Not assigned
1      M2A      Not assigned               Not assigned
2      M3A        North York                  Parkwoods
3      M4A        North York           Victoria Village
4      M5A  Downtown Toronto  Regent Park, Harbourfront

In [4]:
df['Borough'].replace('Not assigned', np.nan, inplace=True)
df.dropna(subset=['Borough'], inplace=True)

df.head(20)

Postcode           Borough  \
2       M3A        North York   
3       M4A        North York   
4       M5A  Downtown Toronto   
5       M6A        North York   
6       M7A  Downtown Toronto   
8       M9A         Etobicoke   
9       M1B       Scarborough   
11      M3B        North York   
12      M4B         East York   
13      M5B  Downtown Toronto   
14      M6B        North York   
17      M9B         Etobicoke   
18      M1C       Scarborough   
20      M3C        North York   
21      M4C         East York   
22      M5C  Downtown Toronto   
23      M6C              York   
26      M9C         Etobicoke   
27      M1E       Scarborough   
30      M4E      East Toronto   

                                        Neighbourhood  
2                                           Parkwoods  
3                                    Victoria Village  
4                           Regent Park, Harbourfront  
5                    Lawrence Manor, Lawrence Heights  
6         Queen's Park, Ontario Provincial Government  
8             Islington Avenue, Humber Valley Village  
9                                      Malvern, Rouge  
11                                          Don Mills  
12                    Parkview Hill, Woodbine Gardens  
13                           Garden District, Ryerson  
14                                          Glencairn  
17  West Deane Park, Princess Gardens, Martin Grov...  
18             Rouge Hill, Port Union, Highland Creek  
20                                          Don Mills  
21                                   Woodbine Heights  
22                                     St. James Town  
23                                 Humewood-Cedarvale  
26  Eringate, Bloordale Gardens, Old Burnhamthorpe...  
27                  Guildwood, Morningside, West Hill  
30                                        The Beaches

In [5]:
df.shape

(103, 3)

# Q2:

In [6]:
df_geo = pd.read_csv('Geospatial_Coordinates.csv')
df_geo.columns = ['Postcode', 'Latitude', 'Longitude']

In [7]:
df_pos = pd.merge(df, df_geo, on=['Postcode'], how='inner')
df_tor = df_pos[['Borough', 'Neighbourhood', 'Postcode', 'Latitude', 'Longitude']].copy()

df_tor.head(10)

Borough                                Neighbourhood Postcode  \
0        North York                                    Parkwoods      M3A   
1        North York                             Victoria Village      M4A   
2  Downtown Toronto                    Regent Park, Harbourfront      M5A   
3        North York             Lawrence Manor, Lawrence Heights      M6A   
4  Downtown Toronto  Queen's Park, Ontario Provincial Government      M7A   
5         Etobicoke      Islington Avenue, Humber Valley Village      M9A   
6       Scarborough                               Malvern, Rouge      M1B   
7        North York                                    Don Mills      M3B   
8         East York              Parkview Hill, Woodbine Gardens      M4B   
9  Downtown Toronto                     Garden District, Ryerson      M5B   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494  
5  43.667856 -79.532242  
6  43.806686 -79.194353  
7  43.745906 -79.352188  
8  43.706397 -79.309937  
9  43.657162 -79.378937

In [8]:
df_pos.head(12)

Postcode           Borough  \
0       M3A        North York   
1       M4A        North York   
2       M5A  Downtown Toronto   
3       M6A        North York   
4       M7A  Downtown Toronto   
5       M9A         Etobicoke   
6       M1B       Scarborough   
7       M3B        North York   
8       M4B         East York   
9       M5B  Downtown Toronto   
10      M6B        North York   
11      M9B         Etobicoke   

                                        Neighbourhood   Latitude  Longitude  
0                                           Parkwoods  43.753259 -79.329656  
1                                    Victoria Village  43.725882 -79.315572  
2                           Regent Park, Harbourfront  43.654260 -79.360636  
3                    Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4         Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
5             Islington Avenue, Humber Valley Village  43.667856 -79.532242  
6                                      Malvern, Rouge  43.806686 -79.194353  
7                                           Don Mills  43.745906 -79.352188  
8                     Parkview Hill, Woodbine Gardens  43.706397 -79.309937  
9                            Garden District, Ryerson  43.657162 -79.378937  
10                                          Glencairn  43.709577 -79.445073  
11  West Deane Park, Princess Gardens, Martin Grov...  43.650943 -79.554724

Answer: We get 103 rows as expected when we do a inner join, so we have good data.